node/src/components/network.rs

```diff
@@ -0,0 +1,610 @@
+mod config;
+mod error;
+mod event;
+mod message;
+
+use std::{
+    collections::{HashMap, HashSet},
+    convert::Infallible,
+    env,
+    fmt::{self, Debug, Display, Formatter},
+    io,
+    marker::PhantomData,
+    num::NonZeroU32,
+    time::Duration,
+};
+
+use datasize::DataSize;
+use futures::{
+    future::{self, BoxFuture, Either},
+    FutureExt,
+};
+use libp2p::{
+    core::{
+        connection::{ConnectedPoint, PendingConnectionError},
+        upgrade,
+    },
+    identity::Keypair,
+    noise::{self, NoiseConfig, X25519Spec},
+    ping::{Ping, PingConfig, PingEvent},
+    swarm::{NetworkBehaviourEventProcess, SwarmBuilder, SwarmEvent},
+    tcp::TokioTcpConfig,
+    yamux::Config as YamuxConfig,
+    Multiaddr, NetworkBehaviour, PeerId, Swarm, Transport,
+};
+use serde::{de::DeserializeOwned, Serialize};
+use tokio::{sync::watch, task::JoinHandle};
+use tracing::{debug, error, info, trace, warn};
+
+use crate::{
+    components::Component,
+    effect::{
+        announcements::NetworkAnnouncement,
+        requests::{NetworkInfoRequest, NetworkRequest},
+        EffectBuilder, EffectExt, Effects,
+    },
+    fatal,
+    reactor::{EventQueueHandle, Finalize, QueueKind},
+    types::NodeId,
+    utils::DisplayIter,
+    NodeRng,
+};
+pub use config::Config;
+pub use error::Error;
+pub(crate) use event::Event;
+pub(crate) use message::Message;
+
+/// The timeout for connection setup (including upgrades) for all inbound and outbound connections.
+// TODO - make this a config option.
+const CONNECTION_TIMEOUT: Duration = Duration::from_secs(10);
+/// Env var which, if it's defined at runtime, enables the libp2p server.
+pub(crate) const ENABLE_LIBP2P_ENV_VAR: &str = "CASPER_ENABLE_LIBP2P";
+
+/// A helper trait whose bounds represent the requirements for a payload that `Network` can
+/// work with.
+pub trait PayloadT:
```

Comment:
I really, really hate these supertraits. They make things unnecessarily complicated and we don't need them, since we never box anything.

Can we not do without these and list bounds explicitly? Often times, we can get away with less bounds, I think.

Reply:
I'd rather leave this as is for this PR in the interests of consistency with other components.

I'm somewhat ambivalent about them personally, but have a mild preference for using them.  I know Ed also likes these, and since I don't believe either of us started the trend, I guess there's at least one more dev in favour of them.

If it's really such a annoyance, let's clean them all up in one separate PR?